include/jemalloc/internal/prof_inlines_b.h

```diff
@@ -123,20 +106,50 @@ prof_sample_accum_update(tsd_t *tsd, size_t usize, bool update,
 		return true;
 	}
 
-	/*
-	 * If this was the first creation of tdata, then
-	 * prof_tdata_get() reset bytes_until_sample, so decrement and
-	 * check it again
-	 */
-	if (!booted && prof_sample_check(tsd, usize, update)) {
-		return true;
+	if (!booted) {
+		/*
+		 * If this was the first creation of tdata, then it means that
+		 * the previous thread_event() relied on the wrong prof_sample
+		 * wait time, and that it should have relied on the new
+		 * prof_sample wait time just set by prof_tdata_get(), so we
+		 * now manually check again.
+		 *
+		 * If the check fails, then even though we relied on the wrong
+		 * prof_sample wait time, we're now actually in perfect shape,
+		 * in the sense that we can pretend that we have used the right
+		 * prof_sample wait time!  The only minor issue is that the
+		 * next prof_sample wait time is identical to the last one,
+		 * rather than a new randomized wait time, but it doesn't
+		 * matter much, and this happens rarely anyway.
+		 *
+		 * If the check succeeds, then we are now in a tougher
+		 * situation, in the sense that we cannot pretend that we have
+		 * used the right prof_sample wait time.  A straightforward
+		 * solution would be to fully roll back thread_event(), set the
+		 * right prof_sample wait time, and then redo thread_event().
+		 * A simpler way, which is implemented below, is to just set a
+		 * new prof_sample wait time that is usize less, and do nothing
+		 * else.  Strictly speaking, the thread event handler may end
+		 * up in a wrong state, since it has still recorded an event
+		 * whereas in reality there may be no event.  However, the
+		 * difference in the wait time offsets the wrongly recorded
+		 * event, so that, functionally, the countdown to the next
+		 * event will behave exactly as if we have used the right
+		 * prof_sample wait time in the first place.
+		 */
+		uint64_t wait = prof_sample_event_wait_get(tsd);
+		assert(wait > 0);
+		if (usize < wait) {
+			thread_prof_sample_event_update(tsd, wait - usize);
+			return true;
+		}
 	}
 
 	if (tsd_reentrancy_level_get(tsd) > 0) {
 		return true;
 	}
 	/* Compute new sample threshold. */
-	if (update) {
+	if (update && booted) {
```

Comment:
Because otherwise we're unnecessarily resetting the threshold twice. See the comments I added in the `!booted` branch.

Reply:
If !booted and we come to this branch, it means we have `usize >= wait`. Shouldn't we consider the initial reset value being consumed already, and reset it again for the next sample? I assume it doesn't hurt to re-init the threshold (it happens only once per thread).

Reply:
Right - it doesn't hurt to reset the threshold again, and in theory the threshold should be a different one since the previous one should be regarded as being consumed already.

I felt it's logically incorrect to immediately overwrite the wait time rather than first calling `thread_event()` to decrease the wait time to `0` and then resetting it (which we couldn't do since that would double count for other events).

Anyway, I've changed it.